### PR TITLE
Revamp module initialization

### PR DESCRIPTION
**Pull request by djmitche**
The diff is pretty formidable, but the individual commits are each pretty bite-sized and tell a nice story.

The overall idea is this: each module registers a `Moduler`, and that `Moduler` gives access to a `Runner` via its `NewRun` method.  In principle, modules could store module-global data (say, a timestamped cache of all installed packages for the `pkg` module) in the `Moduler` object, and should store run-specific data (counts of events, etc.) in the `Runner` object, rather than in global variables :wink:.

### CODE REVIEW

**Comment by jvehent** [comment]
Tests are failing on `src/mig/testutil/modtest.go:15: execution declared and not used`

**Comment by djmitche** [comment]
Sorry it took me so long to loop back to this.. and I don't know how I pushed a change with that bit missing, but hey -- this runs 'make' successfully for me now.

**Comment by djmitche** [comment]
ping?

**Comment by jvehent** [comment]
Ohai from Canterbury, UK. I'm traveling for another couple of weeks, so it may take a while to get eyes on this PR unfortunately :(

On June 4, 2015 6:19:59 PM GMT+01:00, "Dustin J. Mitchell" <notifications@github.com> wrote:
>ping?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/mozilla/mig/pull/64#issuecomment-108980754

-- 
sent from mobile

**Comment by djmitche** [comment]
Oh, right, I totally forgot!

No worries then.  I've got another Go project to tide me over (
https://github.com/djmitche/relengapi-proxy)

Dustin

File: doc/modules.rst [hunk 1]
```diff
@@ -30,20 +30,37 @@ Registration
 A module must import ``mig/modules``.
 
 A module registers itself at runtime via its ``init()`` function which must
-call ``modules.Register`` with a module name and a "runner function" returning
-an anonymous object (``interface {}``) that must implement the
+call ``modules.Register`` with a module name and an object implementing
```
Comment (jvehent) [inline doc/modules.rst]: go doesn't have objects. `modules.Moduler` is an interface. Could you replace `object` with `interface` in that sentence?

File: doc/modules.rst.html [hunk 2]
```diff
@@ -527,4 +527,4 @@
         </section>
     </section>
 </body>
-</html>
\ No newline at end of file
+</html>
```
Comment (jvehent) [inline doc/modules.rst.html]: note to self: this needs to be regenerated with the new content.

File: src/mig/agent/agent.go [hunk 3]
```diff
@@ -163,8 +163,8 @@ func runModuleDirectly(mode string, args []byte, pretty bool) (out string) {
 		return fmt.Sprintf(`{"errors": ["module '%s' is not available"]}`, mode)
 	}
 	// instanciate and call module
-	modRunner := modules.Available[mode].Runner()
-	out = modRunner.(modules.Moduler).Run(os.Stdin)
+	run := modules.Available[mode].NewRun()
```
Comment (jvehent) [inline src/mig/agent/agent.go]: If we're going to rely on `NewRun()` being present, it needs to be part of the `modules.Runner` interface to break compilation if a module forgets to implement it.

```go
type Runner interface {
 	Run(io.Reader) string
 	NewRun() interface{}
 	ValidateParameters() error
 }
```

File: src/mig/client/cmd/main.go [hunk 4]
```diff
@@ -71,7 +71,7 @@ func main() {
 		a                                              mig.Action
 		migrc, show, render, target, expiration, afile string
 		modargs                                        []string
-		modRunner                                      interface{}
+		run                                      interface{}
```
Comment (jvehent) [inline src/mig/client/cmd/main.go]: indent (gofmt)

File: src/mig/agent/agent.go (hunk 3 above)
Comment (jvehent) [inline src/mig/agent/agent.go]: scratch that, it's implemented in the `modules.Moduler` interface.
(really wish github could stage comments...)

**Comment by jvehent** [comment]
It looks good. I'll play a bit more with the interfaces and implement a module that persists data between runs. 
Could you rebase against master? @ameihm0912 added a few tests to the `pkg` module that now break in your branch.
`src/mig/modules/pkg/pkg_test.go:25: undefined: Runner`

File: doc/modules.rst (hunk 1 above)
Comment (djmitche) [inline doc/modules.rst]: Well, interfaces don't implement things.  At least in C, an instance of a struct is idiomatically referred to as an object.  How about "value"

File: src/mig/agent/agent.go (hunk 3 above)
Comment (djmitche) [inline src/mig/agent/agent.go]: Yeah, the name "Moduler" seemed appropriate for values representing a module, where it used to name values representing a run of a module.

File: doc/modules.rst (hunk 1 above)
Comment (djmitche) [inline doc/modules.rst]: Actually, [go does have objects](https://golang.org/doc/faq#Is_Go_an_object-oriented_language).  It's not endowed with all of the behaviors a Java or C++ programmer might expect -- just like a C object -- but the term is still valid

Comment (djmitche) [inline doc/modules.rst]: Ah, how about "instance"?

**Comment by djmitche** [comment]
rebased and fixed.

I'm not sure why travis is failing; it looks like a bug in one of the dependencies:
```
GOPATH=/home/travis/gopath/src/github.com/mozilla/mig GOOS=linux GOARCH=amd64 go get -u golang.org/x/net/icmp
package golang.org/x/net/icmp
	imports golang.org/x/net/internal/iana
	imports golang.org/x/net/internal/iana
	imports golang.org/x/net/internal/iana: use of internal package not allowed
```

**Comment by jvehent** [comment]
Looks like a bug in the `tip` of go. I've updated travis build conf in #69 to use `go 1.4.2` and it passed without error, so if you rebase again we should be good.

**Comment by djmitche** [comment]
re-pushed

**Comment by djmitche** [comment]
Anything else before this gets merged?  It's a high risk of bitrotting..

**Comment by djmitche** [comment]
http://www.gifbin.com/bin/062011/1307380901_time-lapse-strawberry-rotting.gif

**Comment by jvehent** [comment]
Stay put, I'm driving through wonderland!
bc

**Comment by jvehent** [comment]
I finished the review today and reread the updated doc. It all looks good. Merging it.
Thank you Dustin! That's a great contribution. And sorry that it took a while to merge it ;)

**Pull request merged**